### ISSUE TITLE
Model generation does not work in bfloat16

### ISSUE THREAD

**Issue by borisdayma** [borisdayma/dalle-mini#137]
The [inference notebook](https://github.com/borisdayma/dalle-mini/blob/main/tools/inference/inference_pipeline.ipynb) does not work in bfloat16.

It works when setting it to `jnp.float32`.

Here is the error when setting `dtype` from cell 4 of the notebook to `jnp.bfloat16` on a TPU v3-8.

![image](https://user-images.githubusercontent.com/715491/153322423-b1a18c45-1908-40cb-b864-e4fd0ef4eab1.png)

@patil-suraj

**Comment by interactivetech** [comment]
I am seeing the same issue as well

**Comment by pcuenca** [comment]
When attempting to debug this, I noticed a bug: the conversion to `bfloat16` is not actually performed because `model.params` still contains numpy arrays when `model.to_bf16` is invoked, but `jax.tree_map` requires `jnp` arrays. `replicate` takes care of converting to `jnp` arrays, but it is used after the conversion has been attempted. The conversion operation does nothing, but if I fix it, the problem still occurs.

If I initialize the model using `float32` in `from_pretrained`, but convert the parameters to `bf16`, it works. Of course, the computation is still performed in 32-bit mode even if the weights have been downcasted. Therefore, something in the computation itself is triggering this issue.

I tried a few things in an attempt to save memory, but they had no effect whatsoever:
- Initialize the VQGAN and CLIP models after running an initial generation, to force the JIT compiler to run.
- Enable the experimental JIT compiler cache.
- Replace the model.params with empty arrays after they have been replicated.

I'll try to debug it in more depth. Meanwhile, this is how I'm doing the conversion now to fix the issue whereby weights are not being converted:

```python
import numpy as np

def to_jnp(param):
    if isinstance(param, np.ndarray):
        param = jnp.array(param).astype(dtype)
    return param

if dtype == jnp.bfloat16:
    model.params = jax.tree_map(to_jnp, model.params)
```
